src/optionals/queue.c

```diff
@@ -0,0 +1,208 @@
+#include <string.h>
+
+#include "queue.h"
+#include "optionals.h"
+#include "../vm/vm.h"
+
+typedef struct {
+    Value *dq;
+    int capacity;
+    int front;
+    int rear;
+    int count;
+} Queue;
+
+#define AS_QUEUE(v) ((Queue*)AS_ABSTRACT(v)->data)
+
+void freeQueue(DictuVM *vm, ObjAbstract *abstract) {
+    Queue *queue = (Queue*)abstract->data;
+    FREE_ARRAY(vm, Value, queue->dq, queue->capacity);
+    FREE(vm, Queue, abstract->data);
+}
+
+char *queueToString(ObjAbstract *abstract) {
+    UNUSED(abstract);
+
+    char *queueString = malloc(sizeof(char) * 8);
+    snprintf(queueString, 8, "<Queue>");
+    return queueString;
+}
+
+
+static Value queueIsFull(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "size() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    bool isFull = queue->capacity == queue->count;
+
+    return BOOL_VAL(isFull);
+}
+
+static Value queueIsEmpty(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "empty() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    bool isFull = queue->count == 0;
+
+    return BOOL_VAL(isFull);
+}
+
+static Value queueCap(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "cap() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    return NUMBER_VAL(queue->capacity);
+}
+
+static Value queueLen(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "len() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    return NUMBER_VAL(queue->count);
+}
+
+static Value queuePeek(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 0) {
+        runtimeError(vm, "peek() takes no arguments (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+    
+    return newResultSuccess(vm, queue->dq[queue->front]);
+}
+
+static Value queuePush(DictuVM *vm, int argCount, Value *args) {
+    if (argCount != 1) {
+        runtimeError(vm, "push() takes 1 argument (%d given)", argCount);
+        return EMPTY_VAL;
+    }
+
+    Queue *queue = AS_QUEUE(args[0]);
+
+    if (queue->capacity < queue->count + 1) {
+        int oldCapacity = queue->capacity;
+        queue->capacity = GROW_CAPACITY(oldCapacity);
+        queue->dq = GROW_ARRAY(vm, queue->dq, Value, oldCapacity, queue->capacity);
+    }
+
+    if(queue->rear == queue->capacity-1) {
+        queue->rear = -1;            
+    }
+
+    queue->dq[++queue->rear] = args[1];
+    queue->count++;
+
+    return newResultSuccess(vm, NIL_VAL);
```

Comment:
We shouldn't need the Result type here as there isn't an error state

```suggestion
    return NIL_VAL;
```